Formal2024/Workshops/workshop2.lean

```diff
@@ -0,0 +1,239 @@
+import Mathlib.Data.Real.Basic
+import Mathlib.Tactic
+
+namespace Week2
+
+/-
+  # Logic: Implication, universal quantifier, existential quantifier, negation
+  References: [MIL] Mathematics in Lean, [Tut] Tutorials project.
+  Most of the demonstration section comes [MIL].
+-/
+
+/-!
+  # Implication and universal quantifier
+-/
+
+-- To write `∀` type
+
+-- Lean treats the universal quantifier as an implication.
+-- Implications are grouped on the RIGHT (N.B. this is different from the usual convention)
+
+theorem my_lemma : ∀ x y ε : ℝ, 0 < ε → ε ≤ 1 → |x| < ε → |y| < ε → |x * y| < ε :=
+  sorry
+
+-- Lean is all about types, and checking that something is of the right type
+section
+-- Naively, you can think of types as sets. Here these variables are real numbers
+variable (a b δ : ℝ)
+-- Being of the type of a Proposition means being a proof
+variable (h₀ : 0 < δ) (h₁ : δ ≤ 1)
+variable (ha : |a| < δ) (hb : |b| < δ)
+
+-- An implication is something like a function from sets of proofs to set of proofs
+#check my_lemma a b δ
+#check my_lemma a b δ h₀ h₁
+#check my_lemma a b δ h₀ h₁ ha hb
+
+end
+
+
+-- To prove an implication statement, we use intro
+
+example: ∀ x y z : ℝ, x ≤ y → y ≤ z → x ≤ z := by
+  intro x y z h₀ h₁
+  -- Now we can proceed as usual
+  apply le_trans
+  · apply h₀
+  · apply h₁
+
+
+-- Using the universal quantifier, we define an upper bounded function
+def FnUb (f : ℝ → ℝ) (a : ℝ) : Prop :=
+  ∀ x, f x ≤ a
+
+
+
+variable (f g : ℝ → ℝ) (a b : ℝ)
+
+-- It is useful to define functions, this can be done with the command `fun`
+
+#check fun x ↦ x + 1
+#check fun x ↦ f x + g x
+
+-- We can use implication hypothesis to get new hypotheses
+
+example (hfa : FnUb f a) (hgb : FnUb g b) : FnUb (fun x ↦ f x + g x) (a + b) := by
+  intro x
+  dsimp -- This step is needed for lean to unpack the definition of the function
+  apply add_le_add
+  exact hfa x
+  -- `hfa x` is now a proof of `f x ≤ a`
+  exact hgb x
+
+/-
+Since Lean sees an implication as a function from proofs of a Proposition to
+proofs of another Proposition, we can write the proof term explicitly, using `fun`.
+This is more concise, but probably less "human-like". Where to draw the line between
```

Comment:
```suggestion
This is more concise, but probably less "human-like." Where to draw the line between
```

Reply:
I find out now that the convention for this in English is the opposite that the Italian one, fascinating!